extensions/Worker.Extensions.SignalRService/src/Hubs/ServerlessHubOfT.cs

```diff
@@ -0,0 +1,85 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Text.Json;
+using System.Threading.Tasks;
+using Azure.Core.Serialization;
+using Microsoft.AspNetCore.SignalR;
+using Microsoft.Azure.SignalR.Management;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Microsoft.Azure.Functions.Worker.SignalRService
+{
+    public class ServerlessHub<T> where T : class
+    {
+        private static readonly ObjectSerializer JsonObjectSerializer = new JsonObjectSerializer(new(JsonSerializerDefaults.Web));
+        private static readonly NegotiationOptions DefaultNegotiateOptiosn = new();
+
+        private readonly IServiceProvider _serviceProvider;
+
+        private ServiceHubContext<T> HubContext
+        {
+            get
+            {
+                var type = GetType();
+                var hubContextCache = _serviceProvider.GetRequiredService<HubContextCache>();
+                if (hubContextCache.TryGetValue(type, out var hubContext))
+                {
+                    return (ServiceHubContext<T>)hubContext;
+                }
+                var hubContextProviderType = typeof(ServiceHubContextProvider<,>).MakeGenericType(type, typeof(T));
```

Comment:
Curious what the point of this reflection is? Is it to allow for `ServerlessHub<T>` to be derived and still retrieve the exact type from the cache?

I think this can be achieved without reflection, instead since this is all internal and you exclusively control this code, we can rely on returning a lower type and then casting up to the expected type.

``` CSharp
var provider = (ServiceHubContextProvider)_serviceProvider.GetRequiredService(hubContextProviderType);
return (ServiceHubContext<T>)provider.ServiceHubContext;
```

Would then need to also address the lack of common inheritance chain of `ServiceHubContext` and `ServiceHubContext<T>`. Again, since it is all `internal` you could just change the contract to `public object ServiceHubContext { get; }` and rely on all usage to upcast to their expected type.

Reply:
As there might be multiple hubs, there might be multiple `ServiceHubContextProvider` instances in the service provider. The usage of generic parameters of `ServiceHubContextProvider` is to allow multiple such instances coexist in the service provider. Therefore, `ServiceHubContextProvider` has to be generic. However, I think it's still doable to remove the reflection. The updated code always retrieves the hub contexts from `HubContextCache`, whose content are filled by `ServiceHubContextProvider` during the `IHostedService.StartAsync`. 

I have a new question now. Should we make the initialization of `ServiceHubContext` lazy and initialize it when the users first request it? Users don't need to register their serverless hubs with `services.AddServerlessHub`. However, the problem is the initialization of `ServiceHubContext` is async, and it seems that we have to either force the async to be sync or wrap all the properties depending on hub context with `Task<>`. If there are other mechanisms in dotnet worker helping to solve the async problem more elegantly that'd be great.

Reply:
In my suggestion you are still importing the closed-generic type `hubContextProviderType` (this type can be saved as an instance field, so you only need to call `MakeGenericType` once). But the change you made recently addresses the reflection problem well.

As for your other question. Yes, the `IHostedService` approach, and async initialization, is not ideal. But there isn't any single prescribed pattern.  The main drawback of the `IHostedService` approach is if anyone tries to use `ServiceHubContext` before application is fully started, as they may not be initialized yet. If you are fine with that, then you can keep the `IHostedService` approach.

An alternative would be to keep `IHostedService` approach, but then have all the accesses become sync-over-async if they try to access the context before it has been initialized in the background.

Reply:
Make the `IHostedService` `IDisposable`, as `ServiceHubContext` is not `IAsyncDisposable`( The return type of `DisposeAsync` is Task instead of ValueTask due to historical reason).